### PR TITLE
feat(codegen): support calling functions across modules

### PR DESCRIPTION
**Pull request by jaredallard**
Functions are a new concept that are only available to library
templates. A function is a template, declared through `define` that can
be called across modules.

Functions must return a value through the `return` function, which is
only callable inside of a function template. A function template must be
exported through the `module.Export`. Once exported, it can be called
using `module.Call "module-name.FunctionName"`.

Functions behave the same way as templates called through
`ApplyTemplate` do, where data is passed on `.Data` of the `Values`
object.

Example usage of these new template functions:

```go
// module-a
{{- define "HelloWorld" }}
{{- return (printf "Hello, %s!" .Data) }}
{{- end }}
{{- module.Export "HelloWorld" }}

// module-b
{{ module.Call "module-b.HelloWorld" "Jared" }}
// Output: Hello, Jared!
```

### CODE REVIEW

File: internal/codegen/tpl_module.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+//	{{ module.Call "module-b.HelloWorld" "Jared" }}
+//	// Output: Hello, Jared
+func (tm *TplModule) Call(name string, args ...any) (any, error) {
+	// Allows args to not be set.
+	if len(args) > 1 {
+		return nil, fmt.Errorf("Call() only takes max two arguments, name and data")
+	}
+
+	// Functions cannot be called in the first pass.
+	if tm.s.isFirstPass {
```
Comment (deregtd) [inline internal/codegen/tpl_module.go:126]: this seems liek a bit of a dangeorus road to go down.  now you need 3 passes -- first pass to define functions and basics, second to build library calls with function calls, third to actually do the render.

I wonder if you actually want to repeat passes until state stops changing

Comment (deregtd) [reply internal/codegen/tpl_module.go:126]: (and Memoize any native module function calls..?)

File: internal/codegen/tpl_module.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		return nil, nil
+	}
+
+	tm.s.moduleCaller.mu.RLock()
+	defer tm.s.moduleCaller.mu.RUnlock()
+
+	// Get the module name and function name by splitting the name by the
+	// last period.
+	lastPeriodIdx := strings.LastIndex(name, ".")
```
Comment (deregtd) [inline internal/codegen/tpl_module.go:135]: prolly need a validity check here...

File: internal/codegen/tpl_module_test.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			callingTemplate: `{{ module.Call "function.HelloWorld" "hello" }}`,
+			want:            "hello",
+		},
+		{
+			name: "should use data from caller",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return (file.Path) }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ module.Call "function.HelloWorld" }}`,
```
Comment (deregtd) [inline internal/codegen/tpl_module_test.go:82]: add fail case for calling "blah"

File: internal/codegen/tpl_module.go (hunk 1 above)
Comment (jaredallard) [reply internal/codegen/tpl_module.go:126]: We've been doing this two pass system for everything already, module hooks, globals, and applytemplate (which is light version of this).

The only trade off there is that you can't dynamically register functions, which is already the case with Go templates anyways. The way this works out is we just return nothing during that first pass where the second pass is the render results. I haven't seen a case where this doesn't work yet.

File: internal/codegen/tpl_module_test.go (hunk 3 above)
Comment (jaredallard) [reply internal/codegen/tpl_module_test.go:82]: Non-existent? That's the first test case

File: internal/codegen/tpl_module.go (hunk 1 above)
Comment (jaredallard) [reply internal/codegen/tpl_module.go:126]: additional context, VK and I talked about the two pass vs render until stop state and both landed on two pass probably working for everything we'll want to support.

File: internal/codegen/tpl_module.go (hunk 2 above)
Comment (jaredallard) [reply internal/codegen/tpl_module.go:135]: lolz good call

File: internal/codegen/tpl_module_test.go (hunk 3 above)
Comment (jaredallard) [reply internal/codegen/tpl_module_test.go:82]: oic

File: internal/codegen/tpl_module.go (hunk 1 above)
Comment (jaredallard) [reply internal/codegen/tpl_module.go:126]: Will look at changing the state system in the near(ish) future.

**Review by deregtd — APPROVED** [review]

**Pull request merged**